### PR TITLE
[onert] Implement nnfw_train_export_circle (in api)

### PR DESCRIPTION
**Pull request by glistening**
- Introduce OriginIndex in Operand
- Fix to copy the whole operand info during training phase
- Introduce iterateTrainableTensors from api to core
- Introduce _model_path in nnfw_session

ONE-DCO-1.0-Signed-off-by: Sanggyu Lee [sg5.lee@samsung.com](mailto:sg5.lee@samsung.com)

Related: #12223

### CODE REVIEW

**Comment by glistening** [comment]
@Samsung/one_onert Now, it works. I checked using `circle-tensordump` and `onert_train` respectively.

### Run with default param (= epoch 5)

```
$ Product/x86_64-linux.debug/out/bin/onert_train mnist.circle \
--load_input:raw mnist_train.i160.bin \
--load_expected:raw mnist_train.o160.bin

Model Expected Filename mnist_train.o160.bin
Model Input Filename mnist_train.i160.bin
Model Filename mnist.circle
Epoch 1/5 - time: 22.221ms/step - loss: [0] 0.2247
Epoch 2/5 - time: 21.596ms/step - loss: [0] 0.2198
Epoch 3/5 - time: 21.567ms/step - loss: [0] 0.2153
Epoch 4/5 - time: 21.547ms/step - loss: [0] 0.2111
Epoch 5/5 - time: 21.572ms/step - loss: [0] 0.2071
```

### Run with epoch 1

```
$ Product/x86_64-linux.debug/out/bin/onert_train mnist.circle \
--load_input:raw mnist_train.i160.bin \
--load_expected:raw mnist_train.o160.bin \
--epoch 1

Model Expected Filename mnist_train.o160.bin
Model Input Filename mnist_train.i160.bin
Model Filename mnist.circle
Epoch 1/1 - time: 22.565ms/step - loss: [0] 0.2247
```

It generates `tr.circle` which is hard-coded in this DRAFT for test.

```
$ ls mnist.circle tr.circle
-rw-rw-r-- 1 gyu gyu 220948 May 10  2023 mnist.circle
-rw-rw-r-- 1 gyu gyu 220948 Dec 11 16:23 tr.circle

$ md5sum tr.circle mnist.circle
e170cade77fbd5fdbe7c715dc5c91455  tr.circle
6d865d5ed71685854c029d8d8258e6b9  mnist.circle
```

It has same size, but different checksum.

Let's take a closer look into weights with `circle-tensordump`.

```
$ build/compiler/circle-tensordump/circle-tensordump --tensors mnist.circle

[model_3/dense_6/MatMul1]
 ├── shape : (64, 784)
 ├── quantization
 │   ├── min        :
 │   ├── max        :
--
 │   └── quantized_dimension : 0
 └── buffer
     ├── index : 2
     ├── size  : 200704
     └── data  : 0.0150233, 0.0425768, -0.0593695, 0.00335926, -0.0741361, 0.0610657, -0.0489556, -0.0337898,
--
[model_3/dense_7/MatMul]
 ├── shape : (64, 64)
 ├── quantization
 │   ├── min        :
 │   ├── max        :
--
 │   └── quantized_dimension : 0
 └── buffer
     ├── index : 3
     ├── size  : 16384
     └── data  : -0.191558, -0.185649, 0.195834, 0.182175, 0.0182894, -0.00588731, 0.143468, -0.0209067, -0.127524,
--
[model_3/predictions/MatMul]
 ├── shape : (10, 64)
 ├── quantization
 │   ├── min        :
 │   ├── max        :
--
 │   └── quantized_dimension : 0
 └── buffer
     ├── index : 4
     ├── size  : 2560
     └── data  : 0.182483, 0.171852, 0.158546, -0.215122, 0.165991, -0.263981, 0.170174, 0.278153, -0.003059,
```

```
$ build/compiler/circle-tensordump/circle-tensordump --tensors tr.circle

[model_3/dense_6/MatMul1]
 ├── shape : (64, 784)
 ├── quantization
 │   ├── min        :
 │   ├── max        :
--
 │   └── quantized_dimension : 0
 └── buffer
     ├── index : 2
     ├── size  : 200704
     └── data  : 0.0150233, 0.0425768, -0.0593695, 0.00335926, -0.0741361, 0.0610657, -0.0489556, -0.0337898,
--
[model_3/dense_7/MatMul]
 ├── shape : (64, 64)
 ├── quantization
 │   ├── min        :
 │   ├── max        :
--
 │   └── quantized_dimension : 0
 └── buffer
     ├── index : 3
     ├── size  : 16384
     └── data  : -0.191558, -0.185649, 0.195847, 0.182178, 0.0182908, -0.00588296, 0.143469, -0.0208907, -0.127522,
--
[model_3/predictions/MatMul]
 ├── shape : (10, 64)
 ├── quantization
 │   ├── min        :
 │   ├── max        :
--
 │   └── quantized_dimension : 0
 └── buffer
     ├── index : 4
     ├── size  : 2560
     └── data  : 0.18249, 0.171859, 0.158555, -0.214829, 0.166024, -0.263961, 0.170199, 0.278154, -0.00304096,
```
You can see the changes especially from last tensor's buffer.


### Continue training using exported circle
```
$ Product/x86_64-linux.debug/out/bin/onert_train tr.circle \
--load_input:raw mnist_train.i160.bin \
--load_expected:raw mnist_train.o160.bin \
--epoch 4

Model Expected Filename mnist_train.o160.bin
Model Input Filename mnist_train.i160.bin
Model Filename tr.circle
Epoch 1/4 - time: 22.467ms/step - loss: [0] 0.2198
Epoch 2/4 - time: 21.625ms/step - loss: [0] 0.2153
Epoch 3/4 - time: 21.530ms/step - loss: [0] 0.2111
Epoch 4/4 - time: 21.562ms/step - loss: [0] 0.2071
```

Because tr.circle is exported after epoch 1, it resumes and shows the same loss result from epoch 2 ~ epoch 5 from default execution.

Also, I've checked with `mnist_keras.circle`, which has more trainable tensors for `bias`.

I don't put the same repetitious result.

**Comment by glistening** [comment]
It seems failed during building tensorflow under external.

**Comment by glistening** [comment]
@nnfw-bot test tizen-gbs
